### ISSUE TITLE
Mapping from values to functions doesn't work as expected

### ISSUE THREAD

**Issue by funlambda** [fsprojects/Fable#121]
I'm having problems mapping an array of values (like strings) to an array of functions. Here's an example:

```fsharp
let a = [| "a"; "b"; "c" |]

let concaters1 = a |> Array.map (fun x y -> y + x)
let concaters2 = a |> Array.map (fun x -> (fun y -> y + x))
let concaters3 = a |> Array.map (fun x -> let f = (fun y -> y + x) in f)
```

I would expect to get the same value for all three, but only the `concaters3` works correctly. The first two both become `["a0", "b1", "b2"]` in JS, whereas the third is an array of functions (that work as expected). Here's the generated JS:

```js
var concaters1 = a.map(function (x, y) {
  return y + x;
});
var concaters2 = a.map(function (x, y) {
  return y + x;
});
var concaters3 = a.map(function (x) {
  return function () {
    var f;
    return f = function (y) {
      return y + x;
    }, f;
  }();
});
```

Looks like the first two aren't working b/c the generated mapping function is uncurried. The second parameter of the function passed to `Array.prototype.map` is the index, which explains the strange output. Should functions always be generated in curried form in situations like this?

**Comment by alfonsogarciacaro** [comment]
This is different from #120. The problem here comes from the fact that the compiler [is uncurrying the lambdas](https://github.com/fsprojects/Fable/blob/master/src/fable-fsharp/AST/AST.Fable.fs#L436-L454) passed as argument to `fable-core` (under the assumption that this will increase performance) or native methods (for compatibility).

However this is done rather indiscriminately which makes the compiler greedily uncurry some functions like in your case. I'll add an arity check to prevent this.

Thanks a lot for pointing it out!

**Comment by alfonsogarciacaro** [comment]
The solution I came up with was a bit more convoluted than what I'd have liked, but it seems to work :) Both this and #120 fixes have been included in the 0.2.10 release.
